### PR TITLE
Nabla DLL builds for Windows 

### PR DESCRIPTION
**Pull request by AnastaZIuk**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -15,6 +15,23 @@ include(${CMAKE_CURRENT_SOURCE_DIR}/cmake/toolchains/android/build.cmake)
 
 project(Nabla LANGUAGES CXX C)
 
+option(NBL_STATIC_BUILD OFF) # ON for static builds, OFF for shared
```
Comment (devshgraphicsprogramming) [inline CMakeLists.txt:18]: default ON on windows

File: CMakeLists.txt [hunk 2]
```diff
@@ -232,6 +249,12 @@ elseif(UNIX)
 	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -msse4.1")
 endif()
 
+if(NBL_STATIC_BUILD)
+	unset(_NBL_SHARED_BUILD_ CACHE)
+else()
+	set(_NBL_SHARED_BUILD_ ON CACHE INTERNAL "")
+endif()
```
Comment (devshgraphicsprogramming) [inline CMakeLists.txt:256]: what's this?

Comment (AnastaZIuk) [reply CMakeLists.txt:256]: it's for `BuildConfigOptions.h.in`

File: include/nbl/config/BuildConfigOptions.h.in [hunk 3]
```diff
@@ -70,4 +70,16 @@
...
+#cmakedefine _NBL_SHARED_BUILD_
+#ifdef _NBL_SHARED_BUILD_
+#if defined(_NBL_PLATFORM_WINDOWS_) && defined(_MSC_VER)
+#define NBL_EXPORTS __declspec(dllexport)
+#else
+#pragma error("Shared builds are only available for Windows OS & MSVC target for now!")
+#endif
+#else
+#define NBL_EXPORTS
+#endif
```
Comment (devshgraphicsprogramming) [inline include/nbl/config/BuildConfigOptions.h.in:83]: got it backwards, follow my recipe, rename `NBL_EXPORTS` `NBL_API`

NBL_API needs to be `dllexport` when you build nabla, but `dllimport` when you use it

_NBL_SHARED_BUILD_ is nice, it tells me whether I have a static or shared build

```c
#cmakedefine _NBL_SHARED_BUILD_
#ifdef _NBL_SHARED_BUILD_
    #if defined(__NBL_BUILDING_NABLA__) // currently compiling nabla, this define is passed through the commandline
        #if defined(_MSC_VER)
            #define NBL_API __declspec(dllexport)
        #elif defined(__GNUC__)
            #define NBL_API __attribute__ ((visibility ("default")))
        #endif
    #else
        #if defined(_MSC_VER)
            #define NBL_API __declspec(dllimport)
        #else
            #define NBL_API
        #endif
    #endif
#else
    #define NBL_API
#endif
   ```

Comment (AnastaZIuk) [reply include/nbl/config/BuildConfigOptions.h.in]: done

File: include/nbl/asset/IDescriptorSet.h [hunk 4]
```diff
@@ -37,14 +37,16 @@ class NBL_API IDescriptorSet : public virtual core::IReferenceCounted
 
 	public:
 		using layout_t = LayoutType;
-		struct SDescriptorInfo
+		struct NBL_API SDescriptorInfo
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/IDescriptorSet.h:40]: do not export such structs

File: include/nbl/asset/IAccelerationStructure.h [hunk 5]
```diff
@@ -140,6 +140,11 @@ class NBL_API IAccelerationStructure : public IDescriptor
 			E_GEOM_TYPE					type;
 			E_GEOM_FLAGS				flags; // change to core::bitflags later
 			GeometryData<AddressType>	data;
+
+			inline bool operator!=(const Geometry& other) const
+			{
+				return !std::memcmp(this, &other, sizeof(other));
+			}
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/IAccelerationStructure.h:147]: wonder if it all would have worked if you didnt mark `IAccelerationStructure` as wholesale `NBL_API` and just the 3 non-inline methods it has

File: include/nbl/asset/metadata/IAssetMetadata.h [hunk 6]
```diff
@@ -61,7 +61,7 @@ namespace impl{
 	Flexibility has been provided, it is expected each loader has its own base metadata class implementing the 
 	IAssetMetadata interface with its own type enum that other loader's metadata classes derive from the base.
 */
-class NBL_API IAssetMetadata : public impl::IAssetMetadata_base
+class IAssetMetadata : public impl::IAssetMetadata_base
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/metadata/IAssetMetadata.h:64]: you need NBL_API on non-inline methods

File: src/nbl/video/IOpenGLPipeline.h [hunk 7]
```diff
@@ -20,6 +20,13 @@ class IOpenGLPipelineBase
         {
             GLuint GLname = 0u;
             bool uniformsSetForTheVeryFirstTime = true;
+
+            inline bool operator!=(const SShaderProgram& other) const
+            {
+                if (GLname != other.GLname)
+                    return true;
+                return uniformsSetForTheVeryFirstTime != other.uniformsSetForTheVeryFirstTime;
+            }
```
Comment (devshgraphicsprogramming) [inline src/nbl/video/IOpenGLPipeline.h:29]: you shouldn't be having to do anything special for files in `src`

**Comment by AnastaZIuk** [comment]
I think that the idea of inserting `NBL_API` into any class/struct definition located in `nbl/include` was misleading, we should add the keyword to any method of the class or struct that isn't inline and if the entire class is inline - we shouldn't add it anywhere I think. Relying on top level symbol exporting will lead us to many errors.

**Comment by devshgraphicsprogramming** [comment]
> I think that the idea of inserting `NBL_API` into any class/struct definition located in `nbl/include` was misleading, we should add the keyword to any method of the class or struct that isn't inline and if the entire class is inline - we shouldn't add it anywhere I think. Relying on top level symbol exporting will lead us to many errors.

should first attempt to build examples and resolve link errors

**Comment by AnastaZIuk** [comment]
> > I think that the idea of inserting `NBL_API` into any class/struct definition located in `nbl/include` was misleading, we should add the keyword to any method of the class or struct that isn't inline and if the entire class is inline - we shouldn't add it anywhere I think. Relying on top level symbol exporting will lead us to many errors.
> 
> should first attempt to build examples and resolve link errors

right

File: .gitignore [hunk 8]
```diff
@@ -26,6 +26,6 @@ doctemp/*
 source/obj
 source/Irrlicht/libIrrlichtServer.a
 # stuff output by examples
-examples_tests/*/bin/*
+examples_tests/*
```
Comment (devshgraphicsprogramming) [inline .gitignore:29]: shouldn't even be here anymore

Comment (devshgraphicsprogramming) [reply .gitignore:29]: no `examples_tests` path should be here

should be in the submodule's gitignore

File: CMakeLists.txt [hunk 9]
```diff
@@ -15,6 +15,23 @@ include(${CMAKE_CURRENT_SOURCE_DIR}/cmake/toolchains/android/build.cmake)
 
 project(Nabla LANGUAGES CXX C)
 
+option(NBL_STATIC_BUILD "" OFF) # ON for static builds, OFF for shared
```
Comment (devshgraphicsprogramming) [inline CMakeLists.txt:18]: default windows ON

File: src/nbl/CMakeLists.txt [hunk 10]
```diff
@@ -455,6 +455,18 @@ if (ANDROID)
 	)
 endif()
 
+if(NOT NBL_STATIC_BUILD)
```
Comment (devshgraphicsprogramming) [inline src/nbl/CMakeLists.txt:458]: and MSVC toolchain

**Pull request merged**